sdks/python/apache_beam/ml/inference/base.py

```diff
@@ -0,0 +1,194 @@
+import logging
+import platform
+import resource
+import sys
+import time
+from typing import Any
+from typing import List
+
+import apache_beam as beam
+from apache_beam.utils import shared
+from apache_beam.ml.inference.apis import PredictionResult
+
+_MILLISECOND_TO_MICROSECOND = 1000
+_MICROSECOND_TO_NANOSECOND = 1000
+_SECOND_TO_MICROSECOND = 1000000
+
+
+def _unbatch(maybe_keyed_batches: Any):
+  keys, results = maybe_keyed_batches
+  if keys:
+    return zip(keys, results)
+  else:
+    return results
+
+
+class ModelLoader:
+  """Has the ability to load an ML model."""
+  def load_model(self):
+    """Loads an initializes a model for processing."""
+    raise NotImplementedError(type(self))
+
+
+class InferenceRunner:
```

Comment:
Robert had something like 
```
ExampleBatchType = TypeVar('ExampleBatchType')
InferenceBatchType = TypeVar('InferenceBatchType')
 
class InferenceRunner(Generic[ExampleBatchType, InferenceBatchType]):
  def to_impl_batch(batch: Iterable[Any] -> ExampleBatchType):
    raise NotImplementedError(type(self))
 
  def from_impl_batch(inferences: InferenceBatchType) -> Iterable[Any]:
    raise NotImplementedError(type(self))
 
  def run_inference(self, batch : ExampleBatchType) -> InferenceBatchType:
    raise NotImplementedError(type(self))
```
I'm guessing it's not necessary to have the type variables if you're not converting to and from batches? But how do we structure this if we want to leverage Batched DoFn in the future? Thought it's probably not necessary to do the batching explicitly if we can leverage something like `@DoFn.yields_batches`? @TheNeuralBit what are your thoughts?

Reply:
How about doing that change as part of https://issues.apache.org/jira/browse/BEAM-14044. I suppose this sort of feature would depend a lot on how we implement batching more specifically per framework.  Let me add Robert as a reviewer at this point and see what he thinks we should do for batching.